### ISSUE TITLE
sam local invoke does not have response every time

### ISSUE THREAD

**Issue by Li-ReDBox** [aws/aws-sam-cli#3714]
### Description:

Often times, `sam local invoke  HelloWorldFunction` receives output of:
`No response from invoke container for HelloWorldFunction`

### Steps to reproduce:

```
for i in {1..10}
do
   sam local invoke  HelloWorldFunction
   echo; echo
   sleep 5
done
```

### Observed result:

```
for i in {1..10}
> do
>    sam local invoke  HelloWorldFunction
>    echo; echo
>    sleep 5
> done
Invoking app.lambda_handler (python3.9)
Skip pulling image and use local one: public.ecr.aws/sam/emulation-python3.9:rapid-1.40.1-x86_64.

Mounting li-sam-app-py/.aws-sam/build/HelloWorldFunction as /var/task:ro,delegated inside runtime container
No response from invoke container for HelloWorldFunction


Invoking app.lambda_handler (python3.9)
Skip pulling image and use local one: public.ecr.aws/sam/emulation-python3.9:rapid-1.40.1-x86_64.

Mounting li-sam-app-py/.aws-sam/build/HelloWorldFunction as /var/task:ro,delegated inside runtime container
No response from invoke container for HelloWorldFunction


Invoking app.lambda_handler (python3.9)
Skip pulling image and use local one: public.ecr.aws/sam/emulation-python3.9:rapid-1.40.1-x86_64.

Mounting li-sam-app-py/.aws-sam/build/HelloWorldFunction as /var/task:ro,delegated inside runtime container
No response from invoke container for HelloWorldFunction


Invoking app.lambda_handler (python3.9)
Skip pulling image and use local one: public.ecr.aws/sam/emulation-python3.9:rapid-1.40.1-x86_64.

Mounting li-sam-app-py/.aws-sam/build/HelloWorldFunction as /var/task:ro,delegated inside runtime container
No response from invoke container for HelloWorldFunction


Invoking app.lambda_handler (python3.9)
Skip pulling image and use local one: public.ecr.aws/sam/emulation-python3.9:rapid-1.40.1-x86_64.

Mounting li-sam-app-py/.aws-sam/build/HelloWorldFunction as /var/task:ro,delegated inside runtime container
START RequestId: 033c0575-1b6d-4828-bc46-ab35e28d39f4 Version: $LATEST
{}
LambdaContext([aws_request_id=033c0575-1b6d-4828-bc46-ab35e28d39f4,log_group_name=aws/lambda/HelloWorldFunction,log_stream_name=$LATEST,function_name=HelloWorldFunction,memory_limit_in_mb=128,function_version=$LATEST,invoked_function_arn=arn:aws:lambda:us-east-1:012345678912:function:HelloWorldFunction,client_context=None,identity=CognitoIdentity([cognito_identity_id=None,cognito_identity_pool_id=None])])
END RequestId: 033c0575-1b6d-4828-bc46-ab35e28d39f4
REPORT RequestId: 033c0575-1b6d-4828-bc46-ab35e28d39f4	Init Duration: 1.52 ms	Duration: 92.41 ms	Billed Duration: 93 ms	Memory Size: 128 MB	Max Memory Used: 128 MB	
{"statusCode": 200, "body": "{\"message\": \"hello world in build\"}"}

Invoking app.lambda_handler (python3.9)
Skip pulling image and use local one: public.ecr.aws/sam/emulation-python3.9:rapid-1.40.1-x86_64.

Mounting li-sam-app-py/.aws-sam/build/HelloWorldFunction as /var/task:ro,delegated inside runtime container
No response from invoke container for HelloWorldFunction


Invoking app.lambda_handler (python3.9)
Skip pulling image and use local one: public.ecr.aws/sam/emulation-python3.9:rapid-1.40.1-x86_64.

Mounting li-sam-app-py/.aws-sam/build/HelloWorldFunction as /var/task:ro,delegated inside runtime container
No response from invoke container for HelloWorldFunction


Invoking app.lambda_handler (python3.9)
Skip pulling image and use local one: public.ecr.aws/sam/emulation-python3.9:rapid-1.40.1-x86_64.

Mounting li-sam-app-py/.aws-sam/build/HelloWorldFunction as /var/task:ro,delegated inside runtime container
No response from invoke container for HelloWorldFunction


Invoking app.lambda_handler (python3.9)
Skip pulling image and use local one: public.ecr.aws/sam/emulation-python3.9:rapid-1.40.1-x86_64.

Mounting li-sam-app-py/.aws-sam/build/HelloWorldFunction as /var/task:ro,delegated inside runtime container
No response from invoke container for HelloWorldFunction


Invoking app.lambda_handler (python3.9)
Skip pulling image and use local one: public.ecr.aws/sam/emulation-python3.9:rapid-1.40.1-x86_64.

Mounting li-sam-app-py/.aws-sam/build/HelloWorldFunction as /var/task:ro,delegated inside runtime container
No response from invoke container for HelloWorldFunction
```

### Expected result:

To see below output every single time:
```
START RequestId: 033c0575-1b6d-4828-bc46-ab35e28d39f4 Version: $LATEST
{}
LambdaContext([aws_request_id=033c0575-1b6d-4828-bc46-ab35e28d39f4,log_group_name=aws/lambda/HelloWorldFunction,log_stream_name=$LATEST,function_name=HelloWorldFunction,memory_limit_in_mb=128,function_version=$LATEST,invoked_function_arn=arn:aws:lambda:us-east-1:012345678912:function:HelloWorldFunction,client_context=None,identity=CognitoIdentity([cognito_identity_id=None,cognito_identity_pool_id=None])])
END RequestId: 033c0575-1b6d-4828-bc46-ab35e28d39f4
REPORT RequestId: 033c0575-1b6d-4828-bc46-ab35e28d39f4	Init Duration: 1.52 ms	Duration: 92.41 ms	Billed Duration: 93 ms	Memory Size: 128 MB	Max Memory Used: 128 MB	
{"statusCode": 200, "body": "{\"message\": \"hello world in build\"}"}
```

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: MacOS
2. `sam --version`: SAM CLI, version 1.40.1
3. AWS region: local

`Add --debug flag to command you are running`
`--debug` flag does not add any more information.

**Comment by jfuss** [comment]
@Li-ReDBox This usually indicates the timeout is too tight. What happens if you bump the timeout to something a little higher?

**Comment by Li-ReDBox** [comment]
Thanks @jfuss for the suggestions. The `Globals:Functions:Timeout` was 3s as the default, then I have tried 30 and now it is 300, still the majority failed to respond.

When it did run, the invoking did not use a lot of time or resources:
```
REPORT RequestId: c0d1630a-d11c-4493-8d24-95904baf86ac	Init Duration: 3.34 ms	Duration: 102.15 ms	Billed Duration: 103 ms	Memory Size: 128 MB	Max Memory Used: 128 MB	
```
I am using colima, could it be the reason?
```
colima version
colima version v0.3.3
git commit: ac876a5d9e4eda7d1d015896ebfbe1dd16b8c203

runtime: docker
arch: x86_64
client: v20.10.12
server: v20.10.11
```

**Comment by jfuss** [comment]
@Li-ReDBox I am not sure what colima is but looks like a docker replacement? If so, what is causing this is could be solved by https://github.com/aws/aws-sam-cli/pull/3706. If you are willing, you can pull the pr and install that locally to see if that resolves it.

My initial guess here is that our checking for timeout happen once we make an http call to the container, which is not the same thing as the container timing out. This was mostly legacy at this point for reasons I will not bore you with :).

**Comment by Li-ReDBox** [comment]
Hi @jfuss your guess most likely has been confirmed. I run the HelloWorld in a Linux box which has a real Docker, with 3s timeout, there is no error.

Yes, (co)lima is an alternative to Docker for Mac.

It looks like sam needs to mount an image, but it takes time, then it claims: no response. 
```
Mounting /aws/li-sam-app-py/.aws-sam/build/HelloWorldFunction as /var/task:ro,delegated inside runtime container
No response from invoke container for HelloWorldFunction
```
I will check out #3706 to see if helps.  

Thank you.

**Comment by jfuss** [comment]
@Li-ReDBox Awesome!

The PR has some details and I think this resolves it but do let us know if it doesn't as that PR is meant to fix the case you are hitting (or at least that is my understanding of what is going on).

**Comment by Li-ReDBox** [comment]
My findings:

It is wait http response from the container has a very short delay time when [retry](https://github.com/aws/aws-sam-cli/blob/develop/samcli/local/docker/container.py#L275).

In my case, if I change retry to have a longer delay, e.g. 3s instead of the default [delay of 0.05](https://github.com/aws/aws-sam-cli/blob/develop/samcli/lib/utils/retry.py#L9), there is no problem in my test cases:
```python
    @retry(delay=3, exc=requests.exceptions.RequestException, exc_raise=ContainerResponseException)
    def wait_for_http_response(self, name, event, stdout):
        # TODO(sriram-mv): `aws-lambda-rie` is in a mode where the function_name is always "function"
        # NOTE(sriram-mv): There is a connection timeout set on the http call to `aws-lambda-rie`, however there is not
        # a read time out for the response received from the server.
```

**Comment by torresxb1** [comment]
Hi @Li-ReDBox, you're correct in pointing out that increasing the delay of the retries (or even increasing the number of retries) helps. This is because it gives more time for the host port used to communicate with the container to get open. However, the PR mentioned (https://github.com/aws/aws-sam-cli/pull/3706) adds a wait for this port to get open before even  calling this `wait_for_http_response`, so no need to increase retries or delay.

**Comment by torresxb1** [comment]
Closing as this should be addressed in the latest release, 1.47.0.
Please let us know if you still experience issues. Thanks!

**Comment by cdechery** [comment]
I'm still having this same problem, running on Windows with SAM 1.51.

I simply cannot get the "sam invoke" to run the HelloWorld app. It will build but the invoke always times out. I can see the container going up but there seems to be some sort of inability to connect to it and get the Lambda response.
```

2022-09-29 10:55:21,882 | Mounting C:\Users\cdecher\Desktop\Code\Sandbox\lambda-python3.8\.aws-sam\build\HelloWorldFunction as /var/task:ro,delegated inside runtime container
2022-09-29 10:55:23,508 | Starting a timer for 30 seconds for function 'HelloWorldFunction'
2022-09-29 10:55:53,525 | Function 'HelloWorldFunction' timed out after 30 seconds
2022-09-29 10:55:54,681 | Container was not created. Skipping deletion
2022-09-29 10:55:54,681 | Cleaning all decompressed code dirs
2022-09-29 10:55:54,681 | No response from invoke container for HelloWorldFunction
```
